game-core/src/main/java/games/strategy/engine/history/change/units/TransformUnits.java

```diff
@@ -0,0 +1,100 @@
+package games.strategy.engine.history.change.units;
+
+import games.strategy.engine.data.CompositeChange;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.engine.history.change.HistoryChange;
+import games.strategy.triplea.UnitUtils;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.triplea.util.Tuple;
+
+/**
+ * Transforms units into other unit types as determined by {@link
+ * UnitAttachment#getWhenHitPointsDamagedChangesInto()}
+ */
+@Value
+public class TransformUnits implements HistoryChange {
+
+  Territory location;
+  /** Map of old unit -> new unit */
+  Map<Unit, Unit> transformingUnits = new HashMap<>();
+
+  CompositeChange attributeChanges = new CompositeChange();
+
+  public TransformUnits(final Territory location, final Collection<Unit> transformUnits) {
+    this.location = location;
+
+    for (final Unit unit : transformUnits) {
```

Comment:
nit, this could use a quick leading comment to summarize what we are doing in this block, help make it easier to understand.

That could potentially be done by assigning the variable `transformingUnits` in this constructor, in which case the helper method would carry that information, eg:
```
transformingUnits = computeUnitTransformations(transformUnits);
```